### PR TITLE
Fixes Issue - Fig3.5&3.6 (Book1) gauss_plot_2d.ipynb

### PR DESCRIPTION
**Pull request by Aadesh-1404**
| Before PR | After PR |
| :-: | :-: |
|Book1 | New|

### CODE REVIEW

File: gauss_plot_2d.ipynb [hunk 1]
```diff
@@ -0,0 +1,177 @@
+{
```
Comment (patel-zeel) [inline gauss_plot_2d.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #23.    n = 5
</pre><p>Let us use more meaningfull names. n is n_samples? h is height?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/19/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (patel-zeel) [inline gauss_plot_2d.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #48.        ax.plot_surface(X, Y, Z, rstride=3, cstride=3, color='white', edgecolor="gray")
</pre><p>I like the shiny style in the original plot. Can you please use that? maybe check Python data-science handbook for API.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/19/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

**Review by patel-zeel — CHANGES_REQUESTED** [review]
Do not modify the README file.

**Comment by karm-patel** [comment]
Hi @Aadesh-1404, This is great! you fixed some issues which you were facing. I have one comment for 2d figure which was suggested by **Prof. Batra.**

1. Can you add a common color bar for these 3 subfigures? currently, it is not clear which color shows which values, right? you can look this for reference: https://stackoverflow.com/questions/13784201/how-to-have-one-colorbar-for-all-subplots
2. Also, please resolve Zeel's comments, then may be your code will be ready to go in main repo

**Comment by Aadesh-1404** [comment]
Updated as per the comments

| After PR | Before PR |
| :-: | :-: |
| | book3 5|

| After PR | Before PR |
| :-: | :-: |
| | book3 6|


Plot For Contour with Colorbar

3 6_colorbar

**Comment by patel-zeel** [comment]
I have these comments on figures:
3D figure:
* Can you add x, y and z label?

2D figures:
* Size of the title seems larger than expected. Ideally it should match with caption size: 9.
* X and y labels can be added here too. Please add a title for the colorbar.

**Comment by patel-zeel** [comment]
Please refer to the book for exact x,y,z labels to give. For example, Z axis shows pdf and thus we use p(x) to denote that. You can refer to some of the closed PRs as they also use similar convention. 

Ideally, in bivariate normal, x1, x2 will be better but the final decision should be taken as per the book only. Also, labels should be in math mode.

**Comment by patel-zeel** [comment]
Can you set p(x) at this location for the 3d plot? currently it looks bad:
image

Also, as I can see in Eq. 3.18, variable names are used y_1 and y_2 and pdf is a joint pdf so it becomes p(y_1, y_2).

**Comment by patel-zeel** [comment]
Colorbar also shows p(y_1, y_2) so that can be updated too.

**Comment by patel-zeel** [comment]
Also, can you check the updated instructions at `notebooks/README.md` in main repo as well as recently merged notebooks on the main repo to figure out what changes your notebook needs based on the updated instructions?

**Comment by patel-zeel** [comment]
I think we are going with the 2d plot with colorbar, in that case remove the bold titles, keep them unbold. Try to see if you can increase the width as the figure seems compressed a bit in width and there is already some space.

The 3D plot looks good to me.

Also, once you make these changes, you can open a PR on the main repo without the need of further reviews. Please create separate PRs for each figure.

File: gauss_plot_2d.ipynb [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
```
Comment (karm-patel) [inline gauss_plot_2d.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    #### finding multivariate pdf and saving the LATEXIFY figures ####
</pre><p>First split this cell into multiple cells, keep similar code in one cell. for example that two plotting function can be in one cell.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/19/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (karm-patel) [inline gauss_plot_2d.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #13.    class gaussian_pdf(object):
</pre><p>remove this class and just use function written inside this class directly.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/19/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (karm-patel) [inline gauss_plot_2d.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #20.    gauss_pdf = jnp.vectorize(multivariate_pdfs.multi_normal, excluded=[2])
</pre><p>Use <code>jax.vmap()</code>  instead of <code>jnp.vectorize()</code> as suggested by Dr murphy.  Check on jax docs how jax.vmap() works.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/19/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (Aadesh-1404) [reply gauss_plot_2d.ipynb:1]: [jax.numpy.vectorize()](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.vectorize.html#jax.numpy.vectorize) has the same interface as [numpy.vectorize](https://numpy.org/doc/stable/reference/generated/numpy.vectorize.html#numpy.vectorize), but it is syntactic sugar for an auto-batching transformation (vmap()) rather than a Python loop. This should be considerably more efficient, but the implementation must be written in terms of functions that act on JAX arrays.


I have removed the class and directly using the function for vectorization. Do I need to change to Vmap?

Comment (karm-patel) [reply gauss_plot_2d.ipynb:1]: Yes you are right, but JAX uses some lower-level optimization (e.x. memory access optimization) in most methods and perhaps `jax.vmap()` is one of them that's why people used that. So we should also use that ideally.

Comment (Aadesh-1404) [reply gauss_plot_2d.ipynb:1]: I have made the changes suggested. Can you review it once?

File: gauss_plot_2d.ipynb [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
```
Comment (karm-patel) [inline gauss_plot_2d.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #24.        Z = Z = jax.vmap(jax.vmap(gaussian_pdf,(0,0,None)),(1,1,None))(X, Y, gauss)
</pre><p>two times Z</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/19/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (karm-patel) [inline gauss_plot_2d.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #49.        latexify(width_scale_factor=1.5, fig_height=3)
</pre><p>set <code>fig_height=1.5</code> I just checked latex source code. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/19/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (karm-patel) [reply gauss_plot_2d.ipynb:1]: I see, I remembered you told me, okay go ahead with 3.

Comment (karm-patel) [reply gauss_plot_2d.ipynb:1]: But don't change the name of figure when you save, because that name is already referred in latex source code

Comment (karm-patel) [inline gauss_plot_2d.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #20.    savefig('gaussPlot2dDemoContour_Colorbar', tight_layout=False)
</pre><p>Don't add colorbar in figname. keep same figname as previously used.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/19/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (Aadesh-1404) [reply gauss_plot_2d.ipynb:1]: Previously there were three different figures that were created for each gaussian. Now I am creating a subplot so can't have the old name

Comment (Aadesh-1404) [reply gauss_plot_2d.ipynb:1]: The names have changed because previously 
 Gs = ["Full", "Diag", "Spherical"]
 I have changed it to
Gaussians = ["Full", "Diagonal", "Spherical"] for better understanding, apart from that there are no changes

Comment (karm-patel) [inline gauss_plot_2d.ipynb:1]: <p>Add meaningful title in first cell</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/19/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (karm-patel) [inline gauss_plot_2d.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #13.            return multivariate_normal.pdf(jnp.array([x, y]), mean=mean, cov=covariance[G])
</pre><p>There is two tab before <code>return</code> make it one </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/19/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

**Comment by karm-patel** [comment]
After you made changes, go ahead with PR.
